File: Javascript/Arrays/Problem14.js

let hashmap=new Map();

let PrefixSum=0;

let longlen=0;

for(let i=0;i<arr.length;i++){
  
  PrefixSum+=arr[i];
  
  if(PrefixSum==k){
    longlen=Math.max(longlen,i+1);
  }
  
  let rem=PrefixSum-k;
  
  if(hashmap.has(rem)){
    
    let  len=i-hashmap.get(rem);
    
    longlen=Math.max(longlen,len);
    
  }
  if(!hashmap.has(PrefixSum)){
    hashmap.set(PrefixSum,i);
  }
  
}

console.log(longlen)

File: C++/Arrays/Problem14.cpp
#include <iostream>
#include <bits/stdc++.h>
using namespace std;


void SubArr(int n,int arr[],int k){
  
  map<int,int>preSumMap;
  
  int sum=0;
  
  int longlen=0;
  
  for(int i=0;i<n;i++){
    
    sum+=arr[i];
    
    if(sum==k){
      
      longlen=max(longlen,i+1);
    }
    
    int rem=sum-k;
    
    if(preSumMap.find(rem)!=preSumMap.end()){
      int len=i-preSumMap[rem];
      longlen=max(longlen,len);
    }
    
    if(preSumMap.find(sum)==preSumMap.end()){
      
      preSumMap[sum]=i;
    }
    
  }
  std::cout << longlen << std::endl;
}

int main() 
{
    
    int n=6;
    
    int arr[n]={1,2,3,4,5,6};
    
    int k=6;
    
    
    SubArr(n,arr,k);
    
    
    return 0;
}